File: Objective-C Demo App/iOS-SDK-Demo/ALEventTrackingViewController.m
//
//  ALEventTrackingViewController.m
//  iOS-SDK-Demo-ObjC
//
//  Created by Monica Ong on 6/5/17.
//  Copyright © 2017 AppLovin. All rights reserved.
//

#import "ALEventTrackingViewController.h"
#import <StoreKit/StoreKit.h>

#if __has_include(<AppLovinSDK/AppLovinSDK.h>)
    #import <AppLovinSDK/AppLovinSDK.h>
#else
    #import "ALSdk.h"
    #import "ALEventService.h"
#endif

@interface ALDemoEvent : NSObject

@property (nonatomic, copy) NSString *name;
@property (nonatomic, copy) NSString *purpose;

- (instancetype)initWithName:(NSString *)name purpose:(NSString *)purpose;

@end

@interface ALEventTrackingViewController()
@property (nonatomic, strong) NSArray<ALDemoEvent *> *events;
@end

@implementation ALEventTrackingViewController

#pragma mark - View Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
        
    self.events = @[
                    [[ALDemoEvent alloc] initWithName: @"Began Checkout Event" purpose: @"Track when user begins checkout procedure"],
                    [[ALDemoEvent alloc] initWithName: @"Cart Event" purpose: @"Track when user adds an item to cart"],
                    [[ALDemoEvent alloc] initWithName: @"Completed Achievement Event" purpose: @"Track when user completed an achievement"],
                    [[ALDemoEvent alloc] initWithName: @"Completed Checkout Event" purpose: @"Track when user completed checkout"],
                    [[ALDemoEvent alloc] initWithName: @"Completed Level Event" purpose: @"Track when user completed level"],
                    [[ALDemoEvent alloc] initWithName: @"Created Reservation Event" purpose: @"Track when user created a reservation"],
                    [[ALDemoEvent alloc] initWithName: @"In-App Purchase Event" purpose: @"Track when user makes an in-app purchase"],
                    [[ALDemoEvent alloc] initWithName: @"Login Event" purpose: @"Track when user logs in"],
                    [[ALDemoEvent alloc] initWithName: @"Payment Info Event" purpose: @"Tracks when user inputs their payment information"],
                    [[ALDemoEvent alloc] initWithName: @"Registration Event" purpose: @"Track when user registers"],
                    [[ALDemoEvent alloc] initWithName: @"Search Event" purpose: @"Track when user makes a search"],
                    [[ALDemoEvent alloc] initWithName: @"Sent Invitation Event" purpose: @"Track when user sends invitation"],
                    [[ALDemoEvent alloc] initWithName: @"Shared Link Event" purpose: @"Track when user shares a link"],
                    [[ALDemoEvent alloc] initWithName: @"Spent Virtual Currency Event" purpose: @"Track when users spends virtual currency"],
                    [[ALDemoEvent alloc] initWithName: @"Tutorial Event" purpose: @"Track when users does a tutorial"],
                    [[ALDemoEvent alloc] initWithName: @"Viewed Content Event" purpose: @"Track when user views content"],
                    [[ALDemoEvent alloc] initWithName: @"Viewed Product Event" purpose: @"Track when user views product"],
                    [[ALDemoEvent alloc] initWithName: @"Wishlist Event" purpose: @"Track when user adds an item to their wishlist"]
                    ];
}

#pragma mark - Table View Data Source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.events.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *const kCellIdentifier = @"rootPrototype";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier: kCellIdentifier];
    if ( !cell )
    {
        cell = [[UITableViewCell alloc] initWithStyle: UITableViewCellStyleSubtitle reuseIdentifier: kCellIdentifier];
    }
    
    cell.textLabel.text = self.events[indexPath.row].name;
    cell.detailTextLabel.text = self.events[indexPath.row].purpose;
    
    return cell;
}

#pragma mark - Table View Delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [super tableView: tableView didSelectRowAtIndexPath: indexPath];
    
    ALEventService *eventService = [ALSdk shared].eventService;
    
    ALDemoEvent *event = self.events[indexPath.row];
    [self setTitle: event.name];
    
    switch ( indexPath.row )
    {
        case 0:
        {
            [eventService trackEvent: kALEventTypeUserBeganCheckOut
                          parameters: @{kALEventParameterProductIdentifierKey : @"PRODUCT SKU OR ID",
                                        kALEventParameterRevenueAmountKey     : @"PRICE OF ITEM",
                                        kALEventParameterRevenueCurrencyKey   : @"3-LETTER CURRENCY CODE"}];
            break;
        }
        case 1: {
            
            [eventService trackEvent: kALEventTypeUserAddedItemToCart
                          parameters: @{kALEventParameterProductIdentifierKey : @"PRODUCT SKU OR ID"}];
            break;
        }
        case 2:
        {
            [eventService trackEvent: kALEventTypeUserCompletedAchievement
                          parameters: @{kALEventParameterCompletedAchievementKey : @"ACHIEVEMENT NAME OR ID"}];
            break;
        }
        case 3: {
            [eventService trackEvent: kALEventTypeUserCompletedCheckOut
                          parameters: @{kALEventParameterCheckoutTransactionIdentifierKey : @"UNIQUE TRANSACTION ID",
                                        kALEventParameterProductIdentifierKey             : @"PRODUCT SKU OR ID",
                                        kALEventParameterRevenueAmountKey                 : @"AMOUNT OF MONEY SPENT",
                                        kALEventParameterRevenueCurrencyKey               : @"3-LETTER CURRENCY CODE"}];
            break;
        }
        case 4: {
            [eventService trackEvent: kALEventTypeUserCompletedLevel
                          parameters: @{kALEventParameterCompletedLevelKey : @"LEVEL NAME OR NUMBER"}];
            break;
        }
        case 5: {
            [eventService trackEvent: kALEventTypeUserCreatedReservation
                          parameters: @{kALEventParameterProductIdentifierKey    : @"PRODUCT SKU OR ID",
                                        kALEventParameterReservationStartDateKey : @"START NSDATE",
                                        kALEventParameterReservationEndDateKey   : @"END NSDATE"}];
            break;
        }
        case 6: {
            SKPaymentTransaction *transaction = [[SKPaymentTransaction alloc] init]; // from paymentQueue:updatedTransactions:
            //SKProduct* product = ...; // Appropriate product (matching productIdentifier property to SKPaymentTransaction);
            [eventService trackInAppPurchaseWithTransactionIdentifier: transaction.transactionIdentifier
                                                           parameters: @{kALEventParameterRevenueAmountKey : @"AMOUNT OF MONEY SPENT",
                                                                         kALEventParameterRevenueCurrencyKey : @"3-LETTER CURRENCY CODE",
                                                                         kALEventParameterProductIdentifierKey : @"product.productIdentifier"}]; //product.productIdentifier
            break;
        }
        case 7: {
            [eventService trackEvent: kALEventTypeUserLoggedIn
                          parameters: @{kALEventParameterUserAccountIdentifierKey : @"USERNAME"}];
            break;
        }
        case 8: {
            [eventService trackEvent: kALEventTypeUserProvidedPaymentInformation];
            break;
        }
        case 9: {
            [eventService trackEvent: kALEventTypeUserCreatedAccount
                          parameters: @{kALEventParameterUserAccountIdentifierKey : @"USERNAME"}];
            break;
        }
        case 10: {
            [eventService trackEvent: kALEventTypeUserExecutedSearch
                          parameters: @{kALEventParameterSearchQueryKey : @"USER'S SEARCH STRING"}];
            break;
        }
        case 11: {
            [eventService trackEvent: kALEventTypeUserSentInvitation];
            break;
        }
        case 12: {
            [eventService trackEvent: kALEventTypeUserSharedLink];
            break;
        }
        case 13: {
            [eventService trackEvent: kALEventTypeUserSpentVirtualCurrency
                          parameters: @{kALEventParameterVirtualCurrencyAmountKey : @"NUMBER OF COINS SPENT",
                                        kALEventParameterVirtualCurrencyNameKey : @"CURRENCY NAME"}];
            
            break;
        }
        case 14: {
            [eventService trackEvent: kALEventTypeUserCompletedTutorial];
            break;
        }
        case 15: {
            [eventService trackEvent: kALEventTypeUserViewedContent
                          parameters: @{kALEventParameterContentIdentifierKey : @"SOME ID DESCRIBING CONTENT"}];
            break;
        }
        case 16: {
            [eventService trackEvent: kALEventTypeUserViewedProduct
                          parameters: @{kALEventParameterProductIdentifierKey : @"PRODUCT SKU OR ID"}];
            break;
        }
        case 17: {
            [eventService trackEvent: kALEventTypeUserAddedItemToWishlist
                          parameters: @{kALEventParameterProductIdentifierKey : @"PRODUCT SKU OR ID"}];
            break;
        }
        default: {
            [self setTitle: @"Default event tracking initiated"];
            break;
        }
    }
}

@end

@implementation ALDemoEvent

- (instancetype)initWithName:(NSString *)name purpose:(NSString *)purpose
{
    self = [super init];
    if ( self )
    {
        self.name = name;
        self.purpose = purpose;
    }
    return self;
}

@end

File: Swift Demo App/iOS-SDK-Demo/ALEventTrackingViewController.swift
//
//  ALEventTrackingViewController.swift
//  iOS-SDK-Demo-Swift
//
//  Created by Monica Ong on 6/5/17.
//  Copyright © 2017 AppLovin. All rights reserved.
//

import UIKit
import StoreKit
import AppLovinSDK

struct ALDemoEvent
{
    var name: String
    var purpose: String
}

class ALEventTrackingViewController: ALDemoBaseTableViewController
{
    var events: [ALDemoEvent] = []
    
    // MARK: View Lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        events = [ALDemoEvent(name: "Began Checkout Event", purpose: "Track when user begins checkout procedure"),
                  ALDemoEvent(name: "Cart Event", purpose: "Track when user adds an item to cart"),
                  ALDemoEvent(name: "Completed Achievement Event", purpose: "Track when user completed an achievement"),
                  ALDemoEvent(name: "Completed Checkout Event", purpose: "Track when user completed checkout"),
                  ALDemoEvent(name: "Completed Level Event", purpose: "Track when user completed level"),
                  ALDemoEvent(name: "Created Reservation Event", purpose: "Track when user created a reservation"),
                  ALDemoEvent(name: "In-App Purchase Event", purpose: "Track when user makes an in-app purchase"),
                  ALDemoEvent(name: "Login Event", purpose: "Track when user logs in"),
                  ALDemoEvent(name: "Payment Info Event", purpose: "Tracks when user inputs their payment information"),
                  ALDemoEvent(name: "Registration Event", purpose: "Track when user registers"),
                  ALDemoEvent(name: "Search Event", purpose: "Track when user makes a search"),
                  ALDemoEvent(name: "Sent Invitation Event", purpose: "Track when user sends invitation"),
                  ALDemoEvent(name: "Shared Link Event", purpose: "Track when user shares a link"),
                  ALDemoEvent(name: "Spent Virtual Currency Event", purpose: "Track when users spends virtual currency"),
                  ALDemoEvent(name: "Tutorial Event", purpose: "Track when users does a tutorial"),
                  ALDemoEvent(name: "Viewed Content Event", purpose: "Track when user views content"),
                  ALDemoEvent(name: "Viewed Product Event", purpose: "Track when user views product"),
                  ALDemoEvent(name: "Wishlist Event", purpose: "Track when user adds an item to their wishlist")
                 ]
    }
    
    // MARK: Table View Data Source
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return events.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        var cell = tableView.dequeueReusableCell(withIdentifier: "rootPrototype", for: indexPath)
        
        cell.textLabel?.text = events[indexPath.row].name
        cell.detailTextLabel?.text = events[indexPath.row].purpose
        
        return cell
    }
    
    // MARK: Table View Delegate
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath)
    {
        super.tableView(tableView, didSelectRowAt: indexPath)
        
        guard let eventService: ALEventService = ALSdk.shared()?.eventService else {
            title = "Unable to track events because eventService is nil"
            return
        }

        switch indexPath.row {
        case 0:
            title = "User began checkout"
            eventService.trackEvent(kALEventTypeUserBeganCheckOut,
                                    parameters: [kALEventParameterProductIdentifierKey : "PRODUCT SKU OR ID",
                                                 kALEventParameterRevenueAmountKey     : "PRICE OF ITEM",
                                                 kALEventParameterRevenueCurrencyKey   : "3-LETTER CURRENCY CODE"])
        case 1:
            title = "User added item to cart"
            eventService.trackEvent(kALEventTypeUserAddedItemToCart,
                                    parameters: [kALEventParameterProductIdentifierKey : "PRODUCT SKU OR ID"])
            
        case 2:
            title = "User completed achievement"
            eventService.trackEvent(kALEventTypeUserCompletedAchievement,
                                    parameters: [kALEventParameterCompletedAchievementKey : "ACHIEVEMENT NAME OR ID"])
            
        case 3:
            title = "User completed checkout"
            eventService.trackEvent(kALEventTypeUserCompletedCheckOut,
                                    parameters: [kALEventParameterCheckoutTransactionIdentifierKey : "UNIQUE TRANSACTION ID",
                                                 kALEventParameterProductIdentifierKey             : "PRODUCT SKU OR ID",
                                                 kALEventParameterRevenueAmountKey                 : "AMOUNT OF MONEY SPENT",
                                                 kALEventParameterRevenueCurrencyKey               : "3-LETTER CURRENCY CODE"])
        case 4:
            title = "User completed level"
            eventService.trackEvent(kALEventTypeUserCompletedLevel,
                                    parameters: [kALEventParameterCompletedLevelKey : "LEVEL NAME OR NUMBER"])
        case 5:
            title = "User created reservation"
            eventService.trackEvent(kALEventTypeUserCreatedReservation,
                                    parameters: [kALEventParameterProductIdentifierKey    : "PRODUCT SKU OR ID",
                                                 kALEventParameterReservationStartDateKey : "START NSDATE",
                                                 kALEventParameterReservationEndDateKey   : "END NSDATE"])
        case 6:
            title = "Tracking in app purchase"
            //In-App Purchases
            // let transaction: SKPaymentTransaction = ... // from paymentQueue:updatedTransactions:
            //let product: SKProduct = ... // Appropriate product (matching productIdentifier property to SKPaymentTransaction)
            eventService.trackInAppPurchase(withTransactionIdentifier: "transaction.transactionIdentifier",
                                            parameters: [kALEventParameterRevenueAmountKey     : "AMOUNT OF MONEY SPENT",
                                                         kALEventParameterRevenueCurrencyKey   : "3-LETTER CURRENCY CODE",
                                                         kALEventParameterProductIdentifierKey : "product.productIdentifier"]) //product.productIdentifier
        case 7:
            title = "User logged in"
            eventService.trackEvent(kALEventTypeUserLoggedIn,
                                    parameters: [kALEventParameterUserAccountIdentifierKey : "USERNAME"])
        case 8:
            title = "User provided payment information"
            eventService.trackEvent(kALEventTypeUserProvidedPaymentInformation)
            
        case 9:
            title = "User created account"
            eventService.trackEvent(kALEventTypeUserCreatedAccount,
                                    parameters: [kALEventParameterUserAccountIdentifierKey : "USERNAME"])
            
        case 10:
            title = "User executed search"
            eventService.trackEvent(kALEventTypeUserExecutedSearch,
                                    parameters: [kALEventParameterSearchQueryKey : "USER'S SEARCH STRING"])
        case 11:
            title = "User sent invitation"
            eventService.trackEvent(kALEventTypeUserSentInvitation)
        case 12:
            title = "User shared link"
            eventService.trackEvent(kALEventTypeUserSharedLink)
        case 13:
            title = "User spent virtual currency"
            eventService.trackEvent(kALEventTypeUserSpentVirtualCurrency,
                                    parameters: [kALEventParameterVirtualCurrencyAmountKey : "NUMBER OF COINS SPENT",
                                                 kALEventParameterVirtualCurrencyNameKey   : "CURRENCY NAME"])
        case 14:
            title = "User completed tutorial"
            eventService.trackEvent(kALEventTypeUserCompletedTutorial)
        case 15:
            title = "User viewed content"
            eventService.trackEvent(kALEventTypeUserViewedContent,
                                    parameters: [kALEventParameterContentIdentifierKey : "SOME ID DESCRIBING CONTENT"])
        case 16:
            title = "User viewed product"
            eventService.trackEvent(kALEventTypeUserViewedProduct,
                                    parameters: [kALEventParameterProductIdentifierKey : "PRODUCT SKU OR ID"])
        case 17:
            title = "User added item to wishlist"
            eventService.trackEvent(kALEventTypeUserAddedItemToWishlist,
                                    parameters: [kALEventParameterProductIdentifierKey : "PRODUCT SKU OR ID"])
        default:
            title = "Default event tracking initiated"
        }
        
    }
    
}
